multicluster/controllers/multicluster/internal/remote_cluster.go

```diff
@@ -0,0 +1,300 @@
+		config:               config,
+		scheme:               scheme,
+		Namespace:            clusterSetNamespace,
+		connected:            false,
+		isElectedLeader:      false,
+		remoteClusterManager: remoteClusterManager,
+	}
+
+	err = remoteClusterManager.AddRemoteCluster(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	return cluster, nil
+}
+
+func getSecretCACrtAndToken(namespace string, secretName string) ([]byte, []byte, error) {
+	var err error
+	mutex.Lock()
+	if secretCient == nil {
+		secretCient, err = client.New(ctrl.GetConfigOrDie(), client.Options{})
+		if err != nil {
+			mutex.Unlock()
+			return nil, nil, err
+		}
+	}
+	mutex.Unlock()
+	secretObj := &v1.Secret{}
+	secretNamespacedName := types.NamespacedName{
+		Namespace: namespace,
+		Name:      secretName,
+	}
+	err = secretCient.Get(context.TODO(), secretNamespacedName, secretObj)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	caData, found := secretObj.Data[v1.ServiceAccountRootCAKey]
+	if !found {
+		return nil, nil, fmt.Errorf("ca.crt data not found in secret %v", secretName)
+	} else {
+	}
+
+	token, found := secretObj.Data[v1.ServiceAccountTokenKey]
+	if !found {
+		return nil, nil, fmt.Errorf("token not found in secret %v", secretName)
+	}
+
+	return caData, token, nil
+}
+
+func (r *remoteCluster) SendMemberAnnounce() error {
+	log := r.log.WithName("MemberAnnounce" + string(r.ClusterID))
+
+	log.Info("Writing member announce", "cluster", r.GetClusterID())
+
+	memberAnnounceList := &multiclusterv1alpha1.MemberClusterAnnounceList{}
+	if err := r.List(context.TODO(), memberAnnounceList, client.InNamespace(r.GetNamespace())); err != nil {
+		return err
+	}
+	var localClusterMemberAnnounce multiclusterv1alpha1.MemberClusterAnnounce
+	localClusterMemberAnnounceExist := false
+	if len(memberAnnounceList.Items) != 0 {
+		for _, memberAnnounce := range memberAnnounceList.Items {
+			if strings.Compare(memberAnnounce.ClusterID, string(r.remoteClusterManager.GetLocalClusterID())) == 0 {
+				localClusterMemberAnnounceExist = true
+				localClusterMemberAnnounce = memberAnnounce
+			}
+		}
+	}
+	ctx, cancelFunc := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancelFunc()
+	if localClusterMemberAnnounceExist {
+		if r.remoteClusterManager.GetElectedLeaderClusterID() != common.INVALID_CLUSTER_ID {
+			localClusterMemberAnnounce.LeaderClusterID = string(r.remoteClusterManager.GetElectedLeaderClusterID())
+		}
+
+		if localClusterMemberAnnounce.Annotations == nil {
+			localClusterMemberAnnounce.Annotations = make(map[string]string)
+		}
+		localClusterMemberAnnounce.Annotations["touch-ts"] = time.Now().String()
+		// do an update
+		if err := r.Update(ctx, &localClusterMemberAnnounce, &client.UpdateOptions{}); err != nil {
+			log.Error(err, "Failed to update member announce", "cluster", r.GetClusterID())
+			return err
+		}
+	} else {
+		// create
+		localClusterMemberAnnounce.ClusterID = string(r.remoteClusterManager.GetLocalClusterID())
```

Comment:
I am wondering which leader will be for this member cluster? I didn't see you set `LeaderClusterID`.

Reply:
Added a comment